torch/nn/parameter.py

```diff
@@ -44,3 +44,50 @@ def __reduce_ex__(self, proto):
         )
 
     __torch_function__ = _disabled_torch_function_impl
+
+
+class UninitializedParameter(Parameter):
+    r"""A parameter that can be lazily initialized.
+
+    Uninitialized Parameters holds empty tensors that can be moved
+    to different devices and change their type so when they are materialized,
+    they do it directly with the right configuration. Moreover, Unitialized parameters
+    become full fledged :class:`torch.nn.Parameter` after they are initialized.
+    """
+    def __new__(cls, requires_grad=True):
+        data = torch.Tensor()
+        return torch.Tensor._make_subclass(cls, data, requires_grad)
+
+    def materialize(self, shape, device=None, dtype=None):
+        r"""Create a Parameter with the same properties of the uninitialized one.
+        Given a shape, it materializes a parameter in the same device
+        and with the same `dtype` as the current one or the specified ones in the 
+        arguments
+        Args:
+            shape : (tuple): the shape for the materialized tensor.
+            device (:class:`torch.device`): the desired device of the parameters
+                and buffers in this module. Optional.
+            dtype (:class:`torch.dtype`): the desired floating point type of
+                the floating point parameters and buffers in this module. Optional.
+        """
+        if device is None:
+            device = self.data.device
+        if dtype is None:
+            dtype = self.data.dtype
+        self.data = torch.empty(shape, device=device, dtype=dtype)
+        self.__class__ = Parameter
+
+    @property
+    def shape(self):
+        raise RuntimeError(
+            'Can\'t access the shape of an uninitialized parameter. '
```

Comment:
If this is to make sure that loading the state dict of an LazyModule into a regular one is failing, maybe add a note in the error message that if this happens during load_state_dict, this is because you are trying to load an uninitialized parameter into an initialized one.

Reply:
I think it's a generic error message to catch the problems in user code too.

Actually, for completeness - shall we override `size` too? `ndim` I guess is fine as the rank usually doesn't change

Reply:
It is a generic message but also to catch bad loading (from the test, this is the error that is looked for).
I still think it's worth adding a hint in the error if we expect it to be a major reason for this to be raised. Otherwise, people loading models and getting this error will be very confused.